File: experiments/control-run/python/jobq/bench/bench.py
"""The spec's measurements, against `python -m jobq serve` as its own process.

uv run python -m bench.bench [throughput|lag|memory|replay|restart]...
"""

import http.client
import json
import statistics
import subprocess
import sys
import tempfile
import time
from concurrent.futures import ProcessPoolExecutor
from datetime import datetime
from pathlib import Path

from jobq.clock import SystemClock
from jobq.jobs import Job
from jobq.queue import Queue
from jobq.store import LOG_NAME, PutRecord, Store

HOST = "127.0.0.1"
TIMEOUT_S = 60.0  # within: chosen, every bench socket and process wait
THROUGHPUT_SECONDS = 5.0


def write_log(directory: Path, jobs: int, history: str = "queued", lease_until_ms: int = 0) -> int:
    """A log of `jobs` jobs; `history` is queued (1 record), leased (1), or full (4 records)."""
    now = SystemClock().now_ms()
    with (directory / LOG_NAME).open("w") as log:
        for number in range(1, jobs + 1):
            base = Job(
                number=number,
                queue="bench",
                state="queued",
                payload="x" * 100,
                tries=0,
                max_tries=3,
                backoff_ms=0,
                created_ms=now,
                updated_ms=now,
            )
            leased = base.model_copy(
                update={
                    "state": "leased",
                    "tries": 1,
                    "worker": "w",
                    "lease_until_ms": lease_until_ms,
                }
            )
            records = {
                "queued": [base],
                "leased": [leased],
                "full": [
                    base,
                    leased,
                    base.model_copy(update={"tries": 1, "reason": "r"}),
                    base.model_copy(update={"state": "done", "tries": 2, "reason": "r"}),
                ],
            }[history]
            log.writelines(PutRecord(job=job).model_dump_json() + "\n" for job in records)
    return (directory / LOG_NAME).stat().st_size


class Served:
    """`python -m jobq serve <dir> --port 0` in its own process."""

    def __init__(self, directory: Path) -> None:
        started = time.perf_counter()
        self.process = subprocess.Popen(
            [sys.executable, "-m", "jobq", "serve", str(directory), "--port", "0"],
            stderr=subprocess.PIPE,
            text=True,
        )
        assert self.process.stderr is not None
        line = self.process.stderr.readline()
        self.ready_s = time.perf_counter() - started
        self.port = int(line.rsplit(":", 1)[1])

    def rss_mb(self) -> float:
        for line in Path(f"/proc/{self.process.pid}/status").read_text().splitlines():
            if line.startswith("VmRSS:"):
                return int(line.split()[1]) / 1024
        return 0.0

    def stop(self) -> None:
        self.process.terminate()
        self.process.wait(TIMEOUT_S)
        if self.process.stderr is not None:
            self.process.stderr.close()


class Connection:
    """One keep-alive connection for one token."""

    def __init__(self, port: int, token: str) -> None:
        self._http = http.client.HTTPConnection(HOST, port, timeout=TIMEOUT_S)
        self._token = token

    def send(
        self, method: str, path: str, body: dict[str, object] | None = None
    ) -> tuple[int, dict[str, object]]:
        data = None if body is None else json.dumps(body).encode()
        self._http.request(method, path, data, {"authorization": f"Bearer {self._token}"})
        response = self._http.getresponse()
        raw = response.read()
        parsed = json.loads(raw) if raw else {}
        assert isinstance(parsed, dict)
        return response.status, parsed


def _ms(iso: object) -> float:
    assert isinstance(iso, str)
    return datetime.fromisoformat(iso).timestamp() * 1000


def _work(port: int, token: str, seconds: float) -> int:
    connection = Connection(port, token)
    pairs = 0
    end = time.perf_counter() + seconds
    while time.perf_counter() < end:
        status, job = connection.send("POST", "/queues/bench/lease", {"lease_ms": 60_000})
        if status != 200:
            break
        if connection.send("POST", f"/jobs/{job['id']}/ack")[0] == 200:
            pairs += 1
    return pairs


def throughput() -> None:
    for workers in (1, 32):
        with tempfile.TemporaryDirectory() as tmp:
            write_log(Path(tmp), 40_000)
            served = Served(Path(tmp))
            try:
                with ProcessPoolExecutor(workers) as pool:
                    futures = [
                        pool.submit(_work, served.port, f"w{n}", THROUGHPUT_SECONDS)
                        for n in range(workers)
                    ]
                    pairs = sum(f.result(timeout=THROUGHPUT_SECONDS + TIMEOUT_S) for f in futures)
            finally:
                served.stop()
        rate = pairs / THROUGHPUT_SECONDS
        print(f"throughput, {workers} workers: {rate:.0f} lease+ack pairs/s ({2 * rate:.0f} req/s)")


def lag() -> None:
    lease_ms = 300
    with tempfile.TemporaryDirectory() as tmp:
        served = Served(Path(tmp))
        try:
            producer, holder, stream = (Connection(served.port, t) for t in ("p", "a", "b"))
            lags, requests = [], 0
            for _ in range(20):
                body: dict[str, object] = {"queue": "lag", "payload": "x", "max_tries": 100}
                producer.send("POST", "/jobs", body)
                _, leased = holder.send("POST", "/queues/lag/lease", {"lease_ms": lease_ms})
                until = _ms(leased["lease_until"])
                while True:
                    requests += 1
                    status, got = stream.send("POST", "/queues/lag/lease", {"lease_ms": 60_000})
                    if status == 200:
                        lags.append(time.time() * 1000 - until)
                        break
                stream.send("POST", f"/jobs/{got['id']}/ack")
        finally:
            served.stop()
    rate = requests / (20 * lease_ms / 1000)
    print(
        f"lease lag under a steady stream (~{rate:.0f} lease requests/s): "
        f"median {statistics.median(lags):.1f} ms, max {max(lags):.1f} ms over 20 leases"
    )


def memory() -> None:
    for jobs in (0, 100_000):
        with tempfile.TemporaryDirectory() as tmp:
            size = write_log(Path(tmp), jobs)
            served = Served(Path(tmp))
            try:
                Connection(served.port, "m").send("GET", "/health")
                rss = served.rss_mb()
            finally:
                served.stop()
        print(f"resident memory, {jobs} jobs ({size / 1e6:.1f} MB log): {rss:.1f} MB")


def replay() -> None:
    with tempfile.TemporaryDirectory() as tmp:
        size = write_log(Path(tmp), 250_000, history="full")
        started = time.perf_counter()
        store, replayed = Store.open(Path(tmp))
        Queue(store, SystemClock(), replayed)
        elapsed = time.perf_counter() - started
        store.close()
        served = Served(Path(tmp))
        served.stop()
    print(
        f"replay, 1,000,000 records ({size / 1e6:.0f} MB, 250k jobs): {elapsed:.1f} s in process, "
        f"{served.ready_s:.1f} s from process start to listening"
    )


def restart() -> None:
    now = SystemClock().now_ms()
    for label, until in (("still live", now + 3_600_000), ("run out while stopped", now - 1)):
        with tempfile.TemporaryDirectory() as tmp:
            write_log(Path(tmp), 10_000, history="leased", lease_until_ms=until)
            served = Served(Path(tmp))
            try:
                started = time.perf_counter()
                _, health = Connection(served.port, "r").send("GET", "/health")
                first = time.perf_counter() - started
            finally:
                served.stop()
        print(
            f"restart, 10,000 leased jobs, leases {label}: listening in {served.ready_s:.2f} s, "
            f"first request answered {first:.2f} s later, health {health}"
        )


MEASUREMENTS = {
    "throughput": throughput,
    "lag": lag,
    "memory": memory,
    "replay": replay,
    "restart": restart,
}

if __name__ == "__main__":
    for name in sys.argv[1:] or list(MEASUREMENTS):
        MEASUREMENTS[name]()

File: experiments/control-run/elixir/jobq/bench/bench.exs
# The bench client: a real service, real sockets, one keep-alive connection per
# worker. See bench/README.md.

defmodule Bench.Conn do
  @moduledoc false

  def open(port) do
    {:ok, socket} =
      :gen_tcp.connect({127, 0, 0, 1}, port, [:binary, active: false, nodelay: true], 5_000)

    socket
  end

  def request(socket, token, method, path, body \\ nil) do
    body = body || ""

    frame = [
      method,
      " ",
      path,
      " HTTP/1.1\r\nhost: bench\r\nauthorization: Bearer ",
      token,
      "\r\ncontent-length: ",
      Integer.to_string(byte_size(body)),
      "\r\nconnection: keep-alive\r\n\r\n",
      body
    ]

    :ok = :gen_tcp.send(socket, frame)
    :ok = :inet.setopts(socket, packet: :http_bin)
    {:ok, {:http_response, _version, status, _phrase}} = :gen_tcp.recv(socket, 0, 60_000)
    {status, headers(socket, %{})}
  end

  defp headers(socket, acc) do
    case :gen_tcp.recv(socket, 0, 60_000) do
      {:ok, {:http_header, _length, field, _reserved, value}} ->
        headers(socket, Map.put(acc, field |> to_string() |> String.downcase(), value))

      {:ok, :http_eoh} ->
        body(socket, acc)
    end
  end

  defp body(socket, headers) do
    :ok = :inet.setopts(socket, packet: :raw)

    case String.to_integer(Map.get(headers, "content-length", "0")) do
      0 -> ""
      length -> with({:ok, bytes} <- :gen_tcp.recv(socket, length, 60_000), do: bytes)
    end
  end
end

defmodule Bench do
  @moduledoc false

  alias Bench.Conn

  @queue "bench"

  def main(argv) do
    {options, names} = OptionParser.parse!(argv, strict: [jobs: :integer])
    jobs = Keyword.get(options, :jobs, 20_000)
    names = if names == [], do: ~w(throughput expiry silent memory replay restart), else: names

    IO.puts("jobq bench: #{Enum.join(names, ", ")}  (#{jobs} jobs where it matters)")
    IO.puts(String.duplicate("-", 64))
    Enum.each(names, &run(&1, jobs))
  end

  defp run("throughput", jobs) do
    for workers <- [1, 32] do
      %{port: port, stop: stop} = service()
      {microseconds, :ok} = :timer.tc(fn -> fill(port, jobs) end)
      report("create, 8 connections", jobs, microseconds)

      {microseconds, acked} =
        :timer.tc(fn ->
          1..workers
          |> Task.async_stream(fn _n -> work(Conn.open(port), 0) end,
            max_concurrency: workers,
            timeout: :infinity
          )
          |> Enum.reduce(0, fn {:ok, count}, total -> total + count end)
        end)

      report("lease+ack, #{workers} worker(s)", acked, microseconds)
      stop.()
    end
  end

  defp run("expiry", _jobs) do
    %{port: port, stop: stop} = service()
    worker = Conn.open(port)
    {201, _job} = Conn.request(worker, "alice", "POST", "/jobs", job_body(1, 100))
    {200, leased} = Conn.request(worker, "bob", "POST", "/queues/#{@queue}/lease", ~s({"lease_ms":200}))
    until = leased |> JSON.decode!() |> Map.fetch!("lease_until") |> to_unix_ms()

    # A steady stream of lease requests on the queue, which is what the spec
    # asks the lag to be measured under.
    handed_out = poll_until_leased(worker, until)
    IO.puts(pad("lease expiry to hand-out") <> "#{handed_out - until} ms")
    stop.()
  end

  defp run("silent", _jobs) do
    %{port: port, stop: stop} = service()
    before = :erlang.memory(:total)

    silent =
      for _n <- 1..1_200 do
        {:ok, socket} =
          :gen_tcp.connect({127, 0, 0, 1}, port, [:binary, active: false], 5_000)

        socket
      end

    Process.sleep(300)
    held = :erlang.memory(:total) - before

    producer = Conn.open(port)

    times =
      for n <- 1..20 do
        {microseconds, {201, _job}} =
          :timer.tc(fn -> Conn.request(producer, "alice", "POST", "/jobs", job_body(n, 1)) end)

        microseconds
      end
      |> Enum.sort()

    IO.puts(pad("1,200 silent connections") <> "#{mib(held)} MiB on the node")

    IO.puts(
      pad("a create while they hold") <>
        "#{ms(Enum.at(times, 10))} ms median, #{ms(List.first(times))} best, #{ms(List.last(times))} worst, of 20"
    )
    Enum.each(silent, &:gen_tcp.close/1)
    stop.()
  end

  defp run("memory", _jobs) do
    %{port: port, stop: stop, pid: queue} = service()
    fill(port, 100_000)
    :erlang.garbage_collect(queue)
    Process.sleep(200)

    IO.puts(pad("resident memory at 100k jobs") <> "#{rss_mb()} MiB (the node, bench included)")
    IO.puts(pad("BEAM total at 100k jobs") <> "#{mib(:erlang.memory(:total))} MiB")
    IO.puts(pad("the queue process at 100k jobs") <> "#{mib(process_bytes(queue))} MiB")
    stop.()
  end

  defp run("replay", _jobs) do
    dir = tmp_dir()
    records = 1_000_000

    lines =
      Stream.map(1..records, fn n ->
        [
          Jobq.Json.encode(
            Jobq.Job.record(%Jobq.Job{
              n: n,
              queue: @queue,
              payload: "a payload of the usual size, about sixty bytes of it.",
              max_tries: 3,
              tries: rem(n, 3),
              state: :queued,
              created_at: 1_789_000_000_000,
              updated_at: 1_789_000_000_000
            })
          ),
          ?\n
        ]
      end)

    File.open!(Jobq.Store.log_path(dir), [:write, :raw, :binary], fn fd ->
      Enum.each(Stream.chunk_every(lines, 10_000), &:file.write(fd, &1))
    end)

    size = File.stat!(Jobq.Store.log_path(dir)).size
    {microseconds, {:ok, {jobs, _next}}} = :timer.tc(fn -> Jobq.Store.read(dir) end)
    IO.puts(pad("replay of #{records} records") <> "#{ms(microseconds)} ms, #{div(size, 1_048_576)} MiB, #{map_size(jobs)} jobs")

    {microseconds, %{stop: stop}} = :timer.tc(fn -> service(dir) end)
    IO.puts(pad("start on that store") <> "#{ms(microseconds)} ms")
    stop.()
    File.rm_rf(dir)
  end

  defp run("restart", _jobs) do
    dir = tmp_dir()
    %{port: port, stop: stop} = service(dir)
    fill(port, 10_000)
    connection = Conn.open(port)
    Enum.each(1..10_000, fn _n ->
      {200, _job} = Conn.request(connection, "bob", "POST", "/queues/#{@queue}/lease", ~s({"lease_ms":3600000}))
    end)

    stop.()
    {microseconds, %{stop: stop}} = :timer.tc(fn -> service(dir) end)
    IO.puts(pad("start with 10,000 leased jobs") <> "#{ms(microseconds)} ms")
    stop.()
    File.rm_rf(dir)
  end

  defp run(name, _jobs), do: IO.puts("bench: no such measurement: #{name}")

  # The service, and the two ways of talking to it.

  defp service(dir \\ nil) do
    dir = dir || tmp_dir()
    ref = make_ref()
    {:ok, pid} = Jobq.Server.start_link(ref: ref, dir: dir, port: 0)

    %{
      ref: ref,
      pid: Jobq.Registry.whereis(ref, :queue),
      port: Jobq.Server.port(ref),
      dir: dir,
      stop: fn -> Supervisor.stop(pid) end
    }
  end

  defp fill(port, count) do
    connections = for _n <- 1..8, do: Conn.open(port)

    1..count
    |> Enum.chunk_every(div(count, 8) + 1)
    |> Enum.zip(connections)
    |> Task.async_stream(
      fn {chunk, connection} ->
        Enum.each(chunk, fn n ->
          {201, _job} = Conn.request(connection, "alice", "POST", "/jobs", job_body(n, 3))
        end)
      end,
      max_concurrency: 8,
      timeout: :infinity
    )
    |> Stream.run()
  end

  defp work(connection, count) do
    case Conn.request(connection, "bob", "POST", "/queues/#{@queue}/lease", ~s({"lease_ms":60000})) do
      {204, _body} ->
        count

      {200, body} ->
        id = body |> JSON.decode!() |> Map.fetch!("id")
        {200, _acked} = Conn.request(connection, "bob", "POST", "/jobs/#{id}/ack")
        work(connection, count + 1)
    end
  end

  defp poll_until_leased(connection, until) do
    case Conn.request(connection, "eve", "POST", "/queues/#{@queue}/lease", ~s({"lease_ms":1000})) do
      {200, _job} -> System.system_time(:millisecond)
      {204, _body} -> poll_until_leased(connection, until)
    end
  end

  defp job_body(n, max_tries) do
    JSON.encode!(%{
      "queue" => @queue,
      "payload" => "payload number #{n}, about sixty bytes of it all told.",
      "max_tries" => max_tries
    })
  end

  defp to_unix_ms(iso) do
    {:ok, datetime, 0} = DateTime.from_iso8601(iso)
    DateTime.to_unix(datetime, :millisecond)
  end

  defp process_bytes(pid) do
    {:memory, bytes} = Process.info(pid, :memory)
    bytes
  end

  defp rss_mb do
    "/proc/self/status"
    |> File.read!()
    |> String.split("\n")
    |> Enum.find_value(fn line ->
      case String.split(line) do
        ["VmRSS:", kilobytes, _unit] -> div(String.to_integer(kilobytes), 1024)
        _other -> nil
      end
    end)
  end

  defp report(name, count, microseconds) do
    per_second = round(count * 1_000_000 / microseconds)
    IO.puts(pad(name) <> "#{per_second}/s  (#{count} in #{ms(microseconds)} ms)")
  end

  defp ms(microseconds), do: Float.round(microseconds / 1_000, 1)
  defp mib(bytes), do: Float.round(bytes / 1_048_576, 1)
  defp pad(name), do: String.pad_trailing(name <> "  ", 34, ".") <> "  "

  defp tmp_dir do
    # The node's unique integers start over in every VM, so the name carries
    # the OS pid too: two `mix run` invocations must not land on one directory.
    name = "jobq-bench-#{System.pid()}-#{System.unique_integer([:positive])}"
    dir = Path.join(System.tmp_dir!(), name)
    File.mkdir_p!(dir)
    dir
  end
end

Bench.main(System.argv())
